tao_compiler/mlir/disc/transforms/lhlo_legalize_roots_to_loops.cc

```diff
@@ -503,342 +503,546 @@ void emitNotToVectorReduction(OpBuilder& b, Location loc, Operation* root_op,
+// BLOCK TILE LOOP IMPL: block_x * block_y threads load block_x * block_y *
+// tile_h elements, each thread reduces tile_h elements from gmem to SHM(block_x
+// * block_y), do reduction in SHM and atomic to gmem var_tile_h = 64;
+// num_threads_col = 32;
+// num_threads_row = 8;
+// var_threads = num_threads_col * num_threads_row;
+// num_blocks_col = ceil(var_cols / num_threads_col);
+// num_blocks_row = ceil(var_rows / (var_tile_h * num_threads_row));
+// gridDim.y var_blocks = num_blocks_col * num_blocks_row; for (m = 0; m <
+// var_blocks; ++m) {
+//   for (n = 0; n < var_threads; ++n) {
+//     block_col_index = m % num_blocks_col;
+//     block_row_index = m / num_blocks_col;
+//     thread_col_index = n % num_threads_col;
+//     thread_row_index = n / num_threads_col;
+//     shm_trans_index = thread_col_index * num_threads_row + thread_row_index;
+//     col_index = block_col_index * num_threads_col + thread_col_index;
+//     is_col_valid = col_index < var_cols;
+//     if (is_col_valid) {
+//       accum = init_value;
+//       for (int l = 0; l < var_tile_h; ++l) {
+//         row_index = (thread_row_index  + block_row_index * num_threads_row) *
+//         var_tile_h + l;
+//         is_row_valid = row_index < var_rows;
+//         if(is_row_valid) {
+//           accum = accum + global[row_index, col_index];
+//         } else {
+//           accum = accum;
+//         }
+//       }
+//     } else {
+//       accum = init_value;
+//     }
+//     shm[shm_trans_index] = accum;
+//     for (int stride = num_threads_row / 2; stride > 0; stride /= 2) {
+//       __syncthreads();
+//       if (thread_row_index < stride) {
+//         sdata[shm_trans_index] += sdata[shm_trans_index + stride];
+//       }
+//     }
+//     __syncthreads();
+//     if (thread_row_index == 0) {
+//       if (is_col_valid) {
+//         result = sdata[shm_trans_index];
+//         atomicAdd(&global[col_index], result);
+//       }
+//     }
+//   }
+// }
+template <int THREADS_COL, int THREADS_ROW, int TILE_H>
+LogicalResult lowerWithScheduleColReductionTileH(
+    ArrayRef<Operation*> root_ops, Operation* dominant_op, Block* parent,
+    const ShapeAnalysis* shape_analysis = nullptr) {
+  if (!isRank2ColReduction(dominant_op)) {
+    return failure();
   }
 
-  b.setInsertionPointToEnd(for_op_hi.getBody());
-  root_op_idx = 0;
+  // Create helper Values
+  SmallVector<Operation*, 4> col_reduction_roots;
+  std::copy_if(
+      root_ops.begin(), root_ops.end(), std::back_inserter(col_reduction_roots),
+      [](Operation* operation) { return isRank2ColReduction(operation); });
+
+  Value lhs = *dominant_op->getOperands().begin();
+  const int kThreads_col = THREADS_COL;              // 32
+  const int kThreads_row = THREADS_ROW;              // 8
+  const int kTileH = TILE_H;                         // 64
+  const int kTileRow = kThreads_row * TILE_H;        // 512
+  const int kThreads = kThreads_col * kThreads_row;  // 256
```

Comment:
The comments about the value could be removed if they are not fixed.

Reply:
Done.